Add documentation to README (#6)
rancher-sandbox/rancher-desktop-networking: README.md

# rancher-desktop-networking

> Is a simple layer 2 switch that allows connectivity between the VM and the host over `AF_VSOCK` written in Go using [gvisor's](https://github.com/google/gvisor) network stack and [gvisor-tap-vsock](https://github.com/google/gvisor). In addition to piping the ethernet frames, it also provides a configurable DNS, dynamic port forwarding and DHCP.

Reviewer: Please start with a full setence: _`rancher-desktop-networking` is a simple…_ (yes, that means repeating the title; that is fine.)

> Is a simple layer 2 switch that allows connectivity between the VM and the host over `AF_VSOCK` written in Go using [gvisor's](https://github.com/google/gvisor) network stack and [gvisor-tap-vsock](https://github.com/google/gvisor). In addition to piping the ethernet frames, it also provides a configurable DNS, dynamic port forwarding and DHCP.

Reviewer: > … In addition to piping the ethernet frames, it also provides a configurable DNS, dynamic port forwarding and DHCP.

That last part is asymmetrical, and _a… DNS_ doesn't seem to match. Consider something like:
>… In addition to piping the Ethernet frames, it also provides configurable DNS services, DHCP services, and dynamic port forwarding.

Author: It's not a poetry it's the documentation for a layer 2 switch 😂

Rancher Desktop Networking consists of three main components: host-switch, vm-switch, and network-setup. Below is a brief explanation of what each component does.

## host-switch

> For now, it only runs on Windows and acts as a receiving end for all traffic that originated from the WSL VM (designated namespace). It performs a handshake to find the right VM to talk to over `AF_VSOCK`. Once a correct VM is found, it then listens for the incoming traffic from that VM. In additon to this, it can provide a configurable DNS resolver that runs in the user space network along with an API that allows for dynamic port forwarding.

Reviewer: Drop the `For now` and other related verbiage; if it changes in the future, we will be editing the readme at that point.  Consider something like:

> `host-switch` runs on the Windows host and acts as a receiver for all traffic originating from the network namespace within the WSL VM.

> For now, it only runs on Windows and acts as a receiving end for all traffic that originated from the WSL VM (designated namespace). It performs a handshake to find the right VM to talk to over `AF_VSOCK`. Once a correct VM is found, it then listens for the incoming traffic from that VM. In additon to this, it can provide a configurable DNS resolver that runs in the user space network along with an API that allows for dynamic port forwarding.

Reviewer: No `it can provide…` — it's always enabled, and hard coded with our host names. (There's nothing wrong with it being static, but the documentation needs to match the reality.)

## network-setup

> Its main responsibility is to create a network namespace and start the `vm-switch` subprocess in the newly created network namespace.

Reviewer: … running `init` is kind of important, too?

Author: I don't know if init belongs in this documentation, that's more of the Rancher Desktop operation domain. But, if you have any suggestions I'm happy to include them.

Reviewer: `network-setup` still takes an arg for stuff to run under unshare, so we should at least document that part.

## vm-switch

> Once the `vm-switch` process starts in the new namespace, it creates a tap (`eth0`) and a `lo` device. Linux Kernel then forwards all the raw Ethernet frames to the tap device. The tap device forwards the frames over [vsock](https://wiki.qemu.org/Features/VirtioVsock) to the host. The process on the host (host-switch) decapsulates the frames, and since it maintains both internal (vm-switch to host-switch) and external (host-switch to the internet) connections, it uses syscall to connect to the external endpoints.

Reviewer: Something feels off about this…
> Once the `vm-switch` process starts in the new namespace, it creates a tap (`eth0`) and a `lo` device. ~Linux Kernel~ The kernel then forwards all the raw Ethernet frames to the tap device. The tap device forwards the frames over [vsock](https://wiki.qemu.org/Features/VirtioVsock) to the host. The process on the host (~host-switch~ `host-switch`) decapsulates the frames, and ~since it maintains both internal (vm-switch to host-switch) and external (host-switch to the internet) connections, it uses syscall to connect to the external endpoints~ proxies the connections to external hosts via normal syscalls.

Author: I wanted to be more elaborate and explicit on how it works.

> ## Host it works

Reviewer: ```suggestion
## How it works
```

Consider calling this part `Overview` and putting it at the top instead, before we dive into the details on the three parts.

Also consider adding a sequence diagram detailing the handshake process.

```mermaid
flowchart  LR;
 subgraph Host["HOST"]
 subgraph hostSwitch["Host Switch"]

>  vsockHost{"Host Deamon \nListens for Incoming connection"}

Reviewer: Please call this `host-switch.exe` rather than _Host Deamon_ for clarity.  Or maybe you meant this as _part_ of `host-switch.exe`? In which case please don't call it a deamon (a deamon is a process running in the background; a part of a process isn't a process, or even a thread).

 eth(("reconstruct ETH frames"))
 syscall(("OS syscall"))
 dhcp["DHCP"]
 dns["DNS"]
 api["API"]

>  portForwarding["Port Forwarding"]

Reviewer: Once it's constructed, we don't really do anything with port forwarding, right? We just pass it off to `gvisor-tap-vsock`?

Also, if we _do_ want this box, it should be poking out all the way to the host, and not stuck inside this process. (Port forwarding that doesn't interact with the rest of the system is… rather pointless.)

Author: > Once it's constructed, we don't really do anything with port forwarding, right?

The port forwarding runs as an API and you can interact with it via HTTP, socket, etc.

Reviewer: Right, but _we_ (as in, this repo) doesn't do anything with it; `gvisor-tap-vsock` does that. We'll probably do the client end later, but that would be talking to the API box instead.

Reviewer: With more discussion with Nino: we don't do any of DHCP, DNS, API either; so for now we should ignore this and fix it up later.

 vsockHost <----> eth
 eth <----> syscall
 vsockHost ----> dhcp
 vsockHost ----> dns
 vsockHost ----> portForwarding
 vsockHost ----> api
 end
 end
 subgraph VM["VM"]

>  subgraph vmSwitch["VM Switch (Isolated Network Namespace)"]

Reviewer: This is rather confusing. On first reading, I thought this box was just about the namespace itself (and was going to suggest renaming it to just _Isolated network namespace_); however, on re-reading, I guess this is meant to encompass all of `vm-switch` (the process). In that case, please:
- Actually call this `vm-switch`, and _not_ _VM Switch_ — it was vague enough that I thought you meant something else
- Please draw more things in the VM box to make the distinction clear. For example, please draw `init` in a dashed box, and have a box for _Isolated network namespace_ encompassing both `init` and `vm-switch`.

>  vsockVM{"VM Deamon"}

Reviewer: Please don't call a thing that isn't an independent process a _deamon_; that was confusing.

 ethVM(("listens for ETH frames\n from TAP Device"))

>  tapDevice("eth1")

Reviewer: The tap device is now `eth0`, I believe. (It's actually okay to call it `eth1` to make it clear that it's in the namespace, but in either case the documentation needs to match the implementation.)

# rancher-desktop-networking

> `rancher-desktop-networking` is a simple layer 2 switch that allows connectivity between the VM and the host over `AF_VSOCK` written in Go using [gvisor's](https://github.com/google/gvisor) network stack and [gvisor-tap-vsock](https://github.com/google/gvisor). In addition to piping the ethernet frames, it also provides a DNS, DHCP services and dynamic port forwarding.

Reviewer: Count mismatch: _it also provides a_… _services_.
```suggestion
`rancher-desktop-networking` is a simple layer 2 switch that allows connectivity between the VM and the host over `AF_VSOCK` written in Go using [gvisor's](https://github.com/google/gvisor) network stack and [gvisor-tap-vsock](https://github.com/google/gvisor). In addition to piping the ethernet frames, it also provides DNS, DHCP services and dynamic port forwarding.
```

Rancher Desktop Networking consists of three main components: host-switch, vm-switch, and network-setup. Below is a brief explanation of what each component does.

## How it works

> ```mermaid

Reviewer: Please have a comment (at least, as a comment that isn't rendered) describing what the shapes mean — what are rectangles, diamonds, and circles?

flowchart  LR;
 subgraph Host["HOST"]
 subgraph hostSwitch["host-switch.exe"]
 vsockHost{"main loop"}
 eth(("reconstruct ETH frames"))
 syscall(("OS syscall"))
 dhcp["DHCP"]
 dns["DNS"]
 api["API"]
portForwarding["Port Forwarding"]
 vsockHost <----> eth
 eth <----> syscall
 vsockHost ----> dhcp
 vsockHost ----> dns
 vsockHost ----> api
 vsockHost ----> portForwarding
 end
 end
 subgraph VM["VM"]
 subgraph netNs["Isolated Network Namespace"]
 tapDevice("eth0")
 subgraph vmSwitch["vm-switch"]
 vsockVM{"VM Switch"}
 ethVM(("listens for ETH frames\n from TAP Device"))
 ethVM <----> vsockVM
 end
 tapDevice <----> ethVM
 init["/sbin/init"]
 end

>  network-setup["/usr/local/bin/network-setup"]

Reviewer: This (and `init`) is just lying there doing nothing. I guess that's showing its relation with the network namespace, so that's kind of okay?

 end
 vsockVM  <---> |AF_VSOCK| vsockHost
```

## host-switch
`host-switch` runs on the Windows host and acts as a receiver for all traffic originating from the network namespace within the WSL VM. It performs a handshake to find the right VM to talk to over `AF_VSOCK`. Once a correct VM is found, it then listens for the incoming traffic from that VM. In additon to this, it can provide a DNS resolver that runs in the user space network along with an API that allows for dynamic port forwarding.

## network-setup
Its main responsibility is to respond to the handshake request from the `host-switch.exe`, create a network namespace and start the `vm-switch` subprocess in the newly created network namespace. Below is a sequence diagram demonstrating the process.

> ```mermaid

Reviewer: Shouldn't the sequence diagram either be above the description of the individual processes (under the flowchart) or below all three? Having it in the middle seems odd.

sequenceDiagram
    box vm

>     participant script as wsl-init (pid n)

Reviewer: It looks like GitHub's mermaid renderer doesn't like something here… Maybe it doesn't like `box`? https://github.com/mermaid-js/mermaid/issues/4098 indicates it might be relatively new.

    participant network-setup
    participant vm-switch
    participant init as wsl-init (pid 1)
    end
    box host
    participant host-switch as host-switch.exe
    end

    script->>network-setup: spawn
    host-switch->>network-setup: handshake request
    network-setup->>host-switch: handshake response
    host-switch->>network-setup: vsock listener ready
    network-setup->>network-setup: open vsock
    network-setup->>network-setup: create namespace
    network-setup->>vm-switch: spawn
    Note over network-setup,vm-switch: spawn in network namespace, pass in vsock fd
    network-setup->>init: spawn
    Note over network-setup,init: spawns in netns, new mnt/pid ns
    vm-switch->>vm-switch: create lo/eth0
    vm-switch->>vm-switch: DHCP eth0
    vm-switch->>vm-switch: listen for connections
    init-->>init: Spawn /sbin/init
    vm-switch->>host-switch: forward ethernet
```

## vm-switch

> Once the `vm-switch` process starts in the new namespace, it creates a tap (`eth0`) and a `lo` device. The Kernel then forwards all the raw Ethernet frames to the tap device. The tap device forwards the frames over [vsock](https://wiki.qemu.org/Features/VirtioVsock) to the host. The process on the host (`host-switch.exe`) decapsulates the frames, and since it maintains both internal (`vm-switch` to `host-switch.exe`) and external (`host-switch.exe` to the internet) connections, it uses syscall to connect to the external endpoints.

Reviewer: ```suggestion
Once the `vm-switch` process starts in the new namespace, it creates a tap (`eth0`) and a `lo` device. The Kernel then forwards all the raw Ethernet frames to the tap device. The tap device forwards the frames over [vsock](https://wiki.qemu.org/Features/VirtioVsock) to the host. The process on the host (`host-switch.exe`) decapsulates the frames, and since it maintains both internal (`vm-switch` to `host-switch.exe`) and external (`host-switch.exe` to the internet) connections, it connects to the external endpoints via normal syscalls.
```
